app/Services/InboundMail/InboundMailEngine.php

```diff
@@ -0,0 +1,279 @@
+use Log;
+
+class InboundMailEngine
+{
+    use SerializesModels, MakesHash;
+    use GeneratesCounter, SavesDocuments;
+
+    private ?Company $company;
+    private ?bool $isUnknownRecipent = null;
+    private array $globalBlacklistDomains = [];
+    private array $globalBlacklistSenders = [];
+    private array $globalWhitelistDomains = []; // only for global validation, not for allowing to send something into the company, should be used to disabled blocking for mass-senders
+    private array $globalWhitelistSenders = []; // only for global validation, not for allowing to send something into the company, should be used to disabled blocking for mass-senders
+    public function __construct()
+    {
+    }
+    /**
+     * if there is not a company with an matching mailbox, we only do monitoring
+     * reuse this method to add more mail-parsing behaviors
+     */
+    public function handleExpenseMailbox(InboundMail $email)
+    {
+        if ($this->isInvalidOrBlocked($email->from, $email->to))
+            return;
+
+        // Expense Mailbox => will create an expense
+        $company = MultiDB::findAndSetDbByExpenseMailbox($email->to);
+        if (!$company) {
+            $this->saveMeta($email->from, $email->to, true);
+            return;
+        }
+
+        $this->createExpense($company, $email);
+        $this->saveMeta($email->from, $email->to);
+    }
+
+    // SPAM Protection
+    public function isInvalidOrBlocked(string $from, string $to)
+    {
+        // invalid email
+        if (!filter_var($from, FILTER_VALIDATE_EMAIL)) {
+            Log::info('E-Mail blocked, because from e-mail has the wrong format: ' . $from);
+            return true;
+        }
+        if (!filter_var($to, FILTER_VALIDATE_EMAIL)) {
+            Log::info('E-Mail blocked, because to e-mail has the wrong format: ' . $from);
+            return true;
+        }
+
+        $parts = explode('@', $from);
+        $domain = array_pop($parts);
+
+        // global blacklist
+        if (in_array($from, $this->globalWhitelistDomains)) {
+            return false;
+        }
+        if (in_array($domain, $this->globalBlacklistDomains)) {
+            Log::info('E-Mail blocked, because the domain was found on globalBlocklistDomains: ' . $from);
+            return true;
+        }
+        if (in_array($domain, $this->globalWhitelistSenders)) {
+            return false;
+        }
+        if (in_array($from, $this->globalBlacklistSenders)) {
+            Log::info('E-Mail blocked, because the email was found on globalBlocklistEmails: ' . $from);
+            return true;
+        }
+
+        if (Cache::has('inboundMailBlockedSender:' . $from)) { // was marked as blocked before, so we block without any console output
+            return true;
+        }
+
+        // sender occured in more than 500 emails in the last 12 hours
+        $senderMailCountTotal = Cache::get('inboundMailCountSender:' . $from, 0);
+        if ($senderMailCountTotal >= 5000) {
```

Comment:
Should we be storing these thresholds as consts or inside config where they can be accessed / updated via .env var?

Reply:
i started this and will push, when we descided how to move forward with the caching stuff

Reply:
implemented